### PR TITLE
Typescript docs

### PR DESCRIPTION
**Pull request by pettinarip**
## Description

Adding some initial TS docs.

Note: the build is not failing, I canceled it because it was not necessary here.

### CODE REVIEW

File: docs/typescript.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Typescript
+
```
Comment (samajammin) [inline docs/typescript.md:2]: Maybe worth adding a quick intro here, e.g. why we chose to migrate to TS & that we're actively in the process of migrating. e.g.

Our codebase is rapidly increasing in size and in contributors. This brings challenges to reliably keep the code organized, avoid code duplication, and review code. To help alleviate these challenges, we’re in the process of to TypeScript across our codebase. We believe having a strongly typed language will reduce bugs, improve code quality, increase productivity and allow us to scale (both our codebase and our developer community) better in the long term.

File: docs/typescript.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+## General considerations
+
+As we are progressively migrating the entire codebase to TS, we will have a mix of `.js` and `.ts` files in the process. For this reason and to make the migration easier for beginners, we have disabled the [`noImplicitAny`](https://www.typescriptlang.org/tsconfig#noImplicitAny) option from our `tsconfig.json` which means that you can use the `any` type and you will not get an error from that.
+When we reach the moment where the entire codebase is in TS, we will enable `noImplicitAny` and increase the checker’s strictness.
+
+This is just starting, any suggestion is always welcomed and appreciated.
+
+## Migration guide & contribution opportunity 🎉
+
```
Comment (samajammin) [inline docs/typescript.md:11]: I'd add a quick blurb here - "we'd love your help in migrating our codebase!"?

We should specifically point people the issue as a call to action: #6392

File: docs/typescript.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+## GraphQL Typegen
+
+With the release of Gatsby v4.15.0, [the config option `graphqlTypegen` has been added](https://www.gatsbyjs.com/docs/reference/release-notes/v4.15/#graphql-typegen).
+
+- This will generate automatically a `src/gatsby-types.d.ts` file which will contain all the GraphQL query types on the codebase.
+- This will happen in build time. In other words, when you call `yarn start` or `yarn build`.
+- Once you have that file created, a global `Queries` variable will be available which will contain all the query types.
+- More information about how this works https://www.gatsbyjs.com/docs/how-to/local-development/graphql-typegen/
+
+## Other resources
```
Comment (samajammin) [inline docs/typescript.md:88]:
```suggestion
## Learning resources
```

**Comment by pettinarip** [comment]
@samajammin Merging this for now so that I can share it with other people. Let me know in case you need to add something else.

**Pull request merged**